// jquery-file-upload-middleware/index.js
var _ = require('lodash'),
    EventEmitter = require('events').EventEmitter;

var JqueryFileUploadMiddleware = function () {
    EventEmitter.call(this);
    // setting default options
    this.options = this.prepareOptions({});
};
require('util').inherits(JqueryFileUploadMiddleware, EventEmitter);

JqueryFileUploadMiddleware.prototype.prepareOptions = function (options) {
    options = _.extend({
        tmpDir: '/tmp',
        uploadDir: __dirname + '/public/files',
        uploadUrl: '/files/',
        //tianwei
        decoreAppDir: __dirname + '/server/decore',
        decoreDir: __dirname + '/public/storage/decored',
        decoreUrl: '/upload/decore/',
        downloadDir: __dirname + 'public/storage/downloads',
        maxNumberOfFiles: 5,
        maxPostSize: 11000000000, // 11 GB
        minFileSize: 1,
        maxFileSize: 10000000000, // 10 GB
        acceptFileTypes: /.+/i,
        imageTypes: /\.(gif|jpe?g|png)$/i,
        imageVersions: {
//            thumbnail: {
//                width: 80,
//                height: 80
//            }
        },
        accessControl: {
            allowOrigin: '*',
            allowMethods: 'OPTIONS, HEAD, GET, POST, PUT, DELETE'
        }
    }, options);

    //tianwei
    _.each(['uploadDir', 'uploadUrl', 'decoreAppDir', 'decoreDir', 'decoreUrl', 'downloadDir'], function (key) {
        if (!_.isFunction(options[key])) {
            var originalValue = options[key];
            options[key] = function () {
                return originalValue
            };
        }
    });

    return options;
}

JqueryFileUploadMiddleware.prototype.configure = function (options) {
    this.options = this.prepareOptions(options);
};

JqueryFileUploadMiddleware.prototype.fileHandler = function (options) {
    return require('./lib/filehandler')(this, this.prepareOptions(_.extend(this.options, options)));
};

JqueryFileUploadMiddleware.prototype.fileManager = function (options) {
    return require('./lib/filemanager')(this, this.prepareOptions(_.extend(this.options, options)));
};

module.exports = new JqueryFileUploadMiddleware();
