// jquery-file-upload-middleware/lib/fileinfo.js
var fs = require('fs'),
    _ = require('lodash');

module.exports = function (options) {

    var FileInfo = function (file) {
        this.name = file.name;
        this.originalName = file.name;
        this.size = file.size;
        this.type = file.type;
        this.deleteType = 'DELETE';
    };

    FileInfo.prototype.validate = function () {
        if (options.minFileSize && options.minFileSize > this.size) {
            this.error = 'File is too small';
        } else if (options.maxFileSize && options.maxFileSize < this.size) {
            this.error = 'File is too big';
        } else if (!options.acceptFileTypes.test(this.name)) {
            this.error = 'Filetype not allowed';
        } else if (fs.readdirSync(options.baseDir()).length >= options.maxNumberOfFiles) {
            this.error = 'Maximux allowed number of files has been exceeded';
        }
        return !this.error;
    };

    FileInfo.prototype.safeName = function () {
        // Prevent directory traversal and creating hidden system files:
        this.name = require('path').basename(this.name).replace(/^\.+/, '');
        // Prevent overwriting existing files:
        while (fs.existsSync(options.baseDir() + '/' + this.name)) {
            this.name = this.name.replace(/(?:(?: \(([\d]+)\))?(\.[^.]+))?$/, function (s, index, ext) {
                return ' (' + ((parseInt(index, 10) || 0) + 1) + ')' + (ext || '');
            });
        }
    };

    FileInfo.prototype.setUrl = function (type, baseUrl) {
        var key = type ? type + 'Url' : 'url';
        this[key] = baseUrl + '/' + encodeURIComponent(this.name);
    };

    //tianwei
    FileInfo.prototype.setDecoreInfo = function (decore_app, decore_source, decore_dest, decore_url, download_dest) {
        this['decore_url'] = decore_url + '/' + encodeURIComponent(this.name);
        this['decore_app'] = decore_app;
        this['decore_source'] = decore_source + '/' + this.name;
        this['decore_dest'] = decore_dest + '/' + this.name + '.txt';
        this['download_dest'] = download_dest + '/' + this.name + '.cli.txt'
    }

    return FileInfo;
};
